### PR TITLE
DynamicControllerClientBuilder: allows for plugging HTTP wrappers for custom behaviour

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
There might be cases in which callers want to preserve HTTP wrappers for custom behavior per request. For example, there might be distributions that would like to add custom HTTP Headers for each HTTP request. This PR allows for that. 

It uses the `Wrap` method that adds a transport middleware function that will give the caller an opportunity to wrap the underlying `http.RoundTripper` prior to the first API call being made.

We cannot use `WrapTransport` because it simply overwrites RoundTrippers for the current config instead of appending to already existing ones.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/controller-manager/pkg/clientbuilder/client_builder_dynamic.go [hunk 1]
```diff
@@ -235,7 +235,9 @@ func (ts *tokenSourceImpl) Token() (*oauth2.Token, error) {
 
 func constructClient(saNamespace, saName string, config *restclient.Config) restclient.Config {
 	username := apiserverserviceaccount.MakeUsername(saNamespace, saName)
-	ret := *restclient.AnonymousClientConfig(config)
```
Comment (soltysh) [inline staging/src/k8s.io/controller-manager/pkg/clientbuilder/client_builder_dynamic.go:238]: Hmm... we called this to make sure it doesn't hold any identifiable data. Alternative are:
1. Add 2nd method which is similar to `AnonymousClientConfig` but leaves transports
2. Use `AnonymousClientConfig` but copy transport.

Comment (p0lyn0mial) [reply staging/src/k8s.io/controller-manager/pkg/clientbuilder/client_builder_dynamic.go:238]: > Hmm... we called this to make sure it doesn't hold any identifiable data.

yes so this is what we already have for KCM no need to call `AnonymousClientConfig` for each client
https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L223

Comment (soltysh) [reply staging/src/k8s.io/controller-manager/pkg/clientbuilder/client_builder_dynamic.go:238]: Fair.

**Review by soltysh — APPROVED** [review]
/lgtm
@sttts for approval

**Pull request merged**